### ISSUE TITLE
User.roles as an array of objects

### ISSUE THREAD

**Issue by Caroisawesome** [alanning/meteor-roles#86]
Hello,
I am interested in indexing and querying users by groups and/or roles, and the internal structure of user.roles is making it difficult to do so with MongoDB. 
I am having a similar issue to <a href="http://stackoverflow.com/questions/19802502/mongodb-query-help-query-on-values-of-any-key-in-a-sub-object">this person</a>...

My current query for groups is 
```javascript
// user.roles  { group1: ['role1'],  group2: ['role','anotherRole', 'ect'] }

Meteor.users.find({ $where: function() {
    for (var group in this.roles) {
         if (group == 'group1' ) {
              return true;
         }
     }
     return false;
}})
```

If the internal structure consisted of an array of objects like so:

```javascript
user.roles = [
   { group: 'group1',  roles: [ 'role', 'anotherRole'] },
   { group: 'group2', roles: ['admin'] }
]
```
My query for a user with a group could look like this (with a similar query for roles):

```javascript
Meteor.users.find({ roles: {$elemMatch: { group: 'group1' } }})
```

And creating an index with the new structure would look like this:
```javascript
Meteor.users._ensureIndex({ roles.groups: 1 });
```
...where the alternative may not be possible without hard-coded group names.

In conclusion, would you ever consider changing the internal structure of user.roles?
OR is there an easier way to query or index by groups that I am missing? 

Thanks!

**Comment by alanning** [comment]
Thanks for the thoughtful and detailed question!  I especially appreciate you taking the time to research alternatives.

I think your goal is similiar to my own use case and the quick answer is to give each user a separate field that is an array of the 'groups' that they have access to.  For example, in my use case, we partition users by 'network' so we have a `user.networks` field that looks something like this:

```
{
  _id: "abcdefg",
  profile: {
    firstname: "John",
    lastname: "Doe"
  },
  networks: ["00000A", "00000B"],
  roles: {
    "00000A": ["broadcast-alert", "liveview", "checkin"],
    "00000A": ["liveview", "checkin"]
  }
}
```

We create an index on the `networks` field so that we can query users quickly by network.  We can also query by role within a network, like this:

```
var query = {networks: "00000A",
             "roles.00000A": "checkin"}
var usersWithCheckin = db.users.find(query)
```

There are some issues with doing it this way since, as you pointed out, it prevents indexing by group.  It also limits the names of groups (can't start with '$' and have to be careful with periods).  But so far, I've found the limitations to be manageable.

Actually when I first implemented groups, the way you suggested closely matched my first attempt.  The issues that made me switch to the current structure were:

* In our use case, we know the group we want to query and are usually interested in finding users within the group with a specific role.  I don't think MongoDB provides a good way to do a `query by group and role` with the 'array of objects' structure.  Here's a [page with more info](http://stackoverflow.com/questions/3985214/retrieve-only-the-queried-element-in-an-object-array-in-mongodb-collection) on attempting that query.

* Having the index would help when querying for a specific role across groups but the general idea of groups is to partition users so that this would not be very common.

* Using the group as the key makes it a little easier to work with the roles in javascript.  We can just specify the key directly rather than having to search through the array first to find the right group object.

    ```
    var userRoles = user.roles["acme"]

    // vs

    var userRoles2 = _.find(user.roles, function (group) {
      return group === "acme"
    })
    ```

(I think that's how it would work anyways.)


That being said, I would certainly consider changing the internal structure if we could overcome the `query by group and role` issue with the array of objects structure.

**Comment by alanning** [comment]
Doing some further [research](http://docs.mongodb.org/manual/tutorial/query-documents/#specify-multiple-criteria-for-array-elements), it looks like your `$elemMatch` in the query with both a `group` and `role` criteria does achieve the goal:

```
Meteor.users.find({ roles: {$elemMatch: { group: 'group1', role: 'admin' } }})
```

Thanks for pointing that out to me!  I think I do remember seeing that when first adding groups but still went with the current structure for some reason, probably to avoid the `_.find` in the javascript code.

In hindsight, doing it the `$elemMatch` way would avoid the restriction on group names which is probably more important.  Any other limitations you know of to doing it the `$elemMatch` way?

Regardless, its definitely a good candidate for `roles 2.0`.

Regarding `roles 2.0`, my main goal there is to encourage others to be able to better extend `meteor-roles` with their own authorization functionality via their own packages.

For example, there's a couple things that I'd like to see extension packages for:

  * permission hierarchies

  * time-limited access


Any interest in working on `roles 2.0`?

 :-)

**Comment by mitar** [comment]
Schema was changed for v2.0 and is available in its branch. `roles` field is not an array ob objects, but not exactly as it is described in this ticket. Anyway, indexes are now possible.
